message_passing/message_handler_waiting_compensation_cleanup/workflows.py

```diff
@@ -0,0 +1,121 @@
+import asyncio
+from datetime import timedelta
+
+from temporalio import exceptions, workflow
+
+from message_passing.message_handler_waiting_compensation_cleanup import (
+    OnWorkflowExitAction,
+    UpdateInput,
+    WorkflowExitType,
+    WorkflowInput,
+)
+from message_passing.message_handler_waiting_compensation_cleanup.activities import (
+    my_activity,
+)
+
+
+@workflow.defn
+class MyWorkflow:
+    """
+    This Workflow upholds the following recommended practices:
+
+    1. The main workflow method ensures that all signal and update handlers are
+       finished before a successful return, and on failure, cancellation, and
+       continue-as-new.
+    2. The update handler performs any necessary compensation/cleanup when the
+       workflow is cancelled, fails, or continues-as-new.
+    """
+
+    def __init__(self) -> None:
+        self.workflow_exit_exception: asyncio.Future[BaseException] = asyncio.Future()
+
+    @workflow.run
+    async def run(self, input: WorkflowInput) -> str:
+        try:
+            # 👉 Use this `try...except` style, instead of waiting for message
+            # handlers to finish in a `finally` block. The reason is that other
+            # exception types will cause a Workflow Task failure, in which case
+            # we do *not* want to wait for message handlers to finish.
+            result = await self._run(input)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            return result
+        except (
+            asyncio.CancelledError,
+            workflow.ContinueAsNewError,
+            exceptions.FailureError,
```

Comment:
I wonder if we add a wrapper here so that one can do like:
```
except e:
   if is_workflow_exit_exception(e):
     await workflow.wait_condition(workflow.all_handlers_finished)
   raise e
```

Reply:
`asyncio.CancelledError` can occur in many different contexts and should only be a "workflow exit" exception in a small subset of those contexts. I'm not sure about the need for a utility method, and I don't think we would want a function `is_workflow_exit_exception` to simply return `True` for an instance of `asyncio.CancelledError`.

Reply:
> asyncio.CancelledError can occur in many different contexts and should only be considered a "workflow exit" exception in a small subset of those contexts.

It is currently always considered a workflow exit exception when raised from the workflow (see [here](https://github.com/temporalio/sdk-python/blob/2e6d742ce1531741c0f953a311689bd72781b720/temporalio/worker/_workflow_instance.py#L1834-L1836)). It is never treated as a task failure. IMO another way to phrase this helper instead of "is workflow exit exception" is more like "is not workflow task failure exception", but the former reads better as a function name.

Reply:
> I think we may want to make a change to the SDK to make the failure_exception_types set at worker and/or workflow level available.

I think the helper would be enough for now. I don't think a user using those collections directly value (we may change what we consider a workflow exit exception, but yes currently it's Temporal failure, asyncio cancelled error, or an instance of a type in one of those two collections).